### PR TITLE
Adding Factors

### PR DESCRIPTION
**Pull request by omgwtflaserguns**
After the changes the distribution of images delivered by GET are distributed evenly among the crawler sites. This can further be configured by a factor per site between 0.1 and 10, the default value is 1:
```
SoupIO: everyone*2.5
Pr0gramm: static*5.0, new*0.5, top
```
For that to be possible, the crawling had to be changed from crawling to a fixed image count over all crawlers to crawling to a fixed image count per crawler site, as seen in /status:
```
images cached: 185 (2128 bytes) - already crawled: 185 (1680 bytes)
         Reddit - gifs            with factor  1.0: 32 Images [####******]
         Reddit - StarWars        with factor  1.0: 34 Images [####******]*
         Reddit - aww             with factor  1.0: 39 Images [####******]**
         Reddit - FoodPorn        with factor  1.0: 38 Images [####******]**
         Reddit - reactiongifs    with factor  1.0: 38 Images [####******]**
         Reddit - aww_gifs        with factor  1.0:  4 Images [#_________]
```
In the "bar" at the end one character symbols 10% in the cache, the sharps are the part of the cache where crawling is started:
```
[##________]      20% of Cache is filled, crawling will start soon
[####**____]      60% of Cache filled, no crawling expected
[####******]***   130% of Cache filled, crawler added more images then needed
```

### CODE REVIEW

**Comment by jkowalleck** [comment]
@omgwtflaserguns
Why not like this? 

SoupIO: everyone*2.5
Pr0gramm: static*5.0, new*0.5, top

File: config.defaults.ini [hunk 1]
```diff
@@ -12,8 +12,8 @@ IP: 0.0.0.0
 Useragent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10) AppleWebKit/600.1.25 (KHTML, like Gecko) Version/8.0 Safari/600.1.25
 
 [Cache]
```
Comment (jkowalleck) [inline config.defaults.ini]: why did you change this?

File: config.ini [hunk 2]
```diff
@@ -6,8 +6,8 @@ IP: 0.0.0.0
 Useragent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10) AppleWebKit/600.1.25 (KHTML, like Gecko) Version/8.0 Safari/600.1.25
 
 [Cache]
-Images: 150
-Images_min_limit: 20
+Images: 30
```
Comment (jkowalleck) [inline config.ini]: why did you change this?

File: crawler/__init__.py [hunk 3]
```diff
@@ -229,19 +231,66 @@ def __add_image(cls, uri, crawler=None):
...
+                    for site in images[crawler]:
+                        factor = 1
+                        if crawler in factors and site in factors[crawler]:
+                            factor = factors[crawler][site]
+
+                        max += factor
+
+                rnd = random.uniform(0, max)
+
+                count = 0
```
Comment (jkowalleck) [inline crawler/__init__.py]: code doubled. why?

File: crawler/__init__.py [hunk 4]
```diff
@@ -261,16 +314,34 @@ def _debug(cls):
...
         }
 
+        images_count = 0
+        images_size = 0
+        for crawler in images:
+            for site in images[crawler]:
+                site_image_count = len(images[crawler][site])
+                images_count += site_image_count
+                images_size += sys.getsizeof(images[crawler][site])
+                info[crawler + "_" + site] = site_image_count
```
Comment (jkowalleck) [inline crawler/__init__.py]: could we have an own section for this? so the return could look like the following: 
~~~~
{		
  "images" : 23
  "images_size": 1337 ,
  "blacklist":  42 ,
  "blacklist_size":  9000 , 
  "images_sitespecific" : {
    "myCrawler#site" : 2
  } 
 }
~~~~
so it would be easy to distinguish crawlerspecific stats from the general ones and parse them via any script ...

**Comment by jkowalleck** [comment]
@omgwtflaserguns 
still reviewing, most of my comments are markers for myself to remind me of things to check.
Ill leave you a message, when i am done reviewing.

File: crawler/__init__.py [hunk 5]
```diff
@@ -229,19 +231,66 @@ def __add_image(cls, uri, crawler=None):
...
+
+                max = 0
+                for crawler in images:
+                    for site in images[crawler]:
+                        factor = 1
+                        if crawler in factors and site in factors[crawler]:
+                            factor = factors[crawler][site]
+
+                        max += factor
+
```
Comment (jkowalleck) [inline crawler/__init__.py]: calc of `max` might be  improved, will come back later

File: config.defaults.ini (hunk 1 above)
Comment (omgwtflaserguns) [inline config.defaults.ini]: Because the cache size is now per site, so a value of 150 would be a cache of 4200 images

File: crawler/__init__.py (hunk 4 above)
Comment (omgwtflaserguns) [inline crawler/__init__.py]: adressed in bb5ca89
